### PR TITLE
JAMES-2113 Protocols should log to SLF4J

### PR DESCRIPTION
**Pull request by chibenwa**

### CODE REVIEW

File: protocols/api/src/main/java/org/apache/james/protocols/api/logger/ContextualLogger.java [hunk 1]
```diff
@@ -42,7 +44,7 @@ public String get() {
         }, session.getSessionID(), logger);
     }
 
-    public ContextualLogger(Supplier<String> userSupplier, String sessionId, Logger logger) {
+    public ContextualLogger(Supplier<String> userSupplier, String sessionId, org.slf4j.Logger logger) {
```
Comment (mbaechler) [inline protocols/api/src/main/java/org/apache/james/protocols/api/logger/ContextualLogger.java]: Ok, so the full concept of ContextualLogger is no longer relevant : we should use MDC for that (see slf4j and/or logback MDC for more information).
As MDC are ThreadLocal, we have to solve thread + MDC in our context.

Comment (chibenwa) [inline protocols/api/src/main/java/org/apache/james/protocols/api/logger/ContextualLogger.java]: IMO we should not couple this.

MDC is of course desirable, but I prefer working on it separatly:
 - Java-8 PR would allow auto-closable
 - No need to delay a bugfix
 - It seems like an unrelated issue that deserves it's own ticket.

Comment (chibenwa) [inline protocols/api/src/main/java/org/apache/james/protocols/api/logger/ContextualLogger.java]: Please see https://issues.apache.org/jira/browse/JAMES-2114

**Comment by chibenwa** [comment]
see for the MDC switch and a deeper migration to SLF4J https://github.com/linagora/james-project/pull/928

File: server/protocols/protocols-imap4/src/main/java/org/apache/james/imapserver/netty/ImapChannelUpstreamHandler.java [hunk 2]
```diff
@@ -98,7 +98,7 @@ public ImapChannelUpstreamHandler(String hello, ImapProcessor processor, ImapEnc
     private Logger getLogger(final ChannelHandlerContext ctx) {
         return new ContextualLogger(
                 getUserSupplier(ctx),
-                "" + ctx.getChannel().getId(),
+                Integer.toString(ctx.getChannel().getId()),
```
Comment (mbaechler) [inline server/protocols/protocols-imap4/src/main/java/org/apache/james/imapserver/netty/ImapChannelUpstreamHandler.java]: you should rather use String.valueOf (yes, it's stupid Java things ...)

File: server/protocols/protocols-imap4/src/main/java/org/apache/james/imapserver/netty/ImapChannelUpstreamHandler.java [hunk 3]
```diff
@@ -95,22 +96,41 @@ public ImapChannelUpstreamHandler(String hello, ImapProcessor processor, ImapEnc
...
+    }
+
+    private Supplier<String> getUserSupplier(final ChannelHandlerContext ctx) {
+        return () -> {
+            Object o = attributes.get(ctx.getChannel());
+            if (o instanceof IMAPSession) {
+                IMAPSession session = (IMAPSession) o;
+                return session.getUser();
+            }
+            return null;
```
Comment (quynhn) [inline server/protocols/protocols-imap4/src/main/java/org/apache/james/imapserver/netty/ImapChannelUpstreamHandler.java]: We don't like "null", do we? We prefer Optional?

Comment (chibenwa) [inline server/protocols/protocols-imap4/src/main/java/org/apache/james/imapserver/netty/ImapChannelUpstreamHandler.java]: ContextualLogger is anyway removed in https://github.com/linagora/james-project/pull/928

**Comment by chibenwa** [comment]
Merged
